diff --git a/httpstream.go b/httpstream.go
index 2981bf1..96d5826 100644
--- a/httpstream.go
+++ b/httpstream.go
@@ -78,6 +78,7 @@ func (h *httpStreamFactory) New(net, transport gopacket.Flow) tcpassembly.Stream
 		transport: transport,
 		r:         tcpreader.NewReaderStream(),
 	}
+	h.wg.Add(1)
 	go hstream.run(h.wg) // Important... we must guarantee that data from the reader stream is read.
 
 	// ReaderStream implements tcpassembly.Stream, so we can return a pointer to it.
@@ -85,6 +86,7 @@ func (h *httpStreamFactory) New(net, transport gopacket.Flow) tcpassembly.Stream
 }
 
 func (h *httpStream) run(wg *sync.WaitGroup) {
+	defer wg.Done()
 	buf := bufio.NewReader(&h.r)
 
 	src := h.net.Src().String()
diff --git a/main.go b/main.go
index 4e3dcaf..fdad2d6 100644
--- a/main.go
+++ b/main.go
@@ -116,6 +116,7 @@ func readStream(wg *sync.WaitGroup, r *pcapgo.Reader) {
 		case packet := <-packets:
 			// A nil packet indicates the end of a pcap file.
 			if packet == nil {
+				assembler.FlushAll()
 				return
 			}
 
